### PR TITLE
Create C++ Custom Visualizer sample

### PR DESCRIPTION
**Pull request by gregg-miskelly**
This commit provides an example C++ custom visualizer. See /CppCustomVisualizer/README.md for more information.

### CODE REVIEW

File: CppCustomVisualizer/README.md [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+the debugger based on the type of that variable (example: change the way CExampleClass variables
+are displayed) using .natvis files. Natvis files add rules to format a type either declaratively or
+by running custom code in the debugger. This custom code is called a Custom Visualizer. Most of
+the time as it easier and much less error prone to use the declarative approach. But
+Custom Visualizers are useful when formatting a type is too complicated to be done declaratively.
+
+Natvis documentation can be found on [docs.microsoft.com](https://docs.microsoft.com/en-us/visualstudio/debugger/create-custom-views-of-native-objects).
+
+## How to use this sample
+More information about this sample can be found in the [Wiki for this project](https://github.com/Microsoft/ConcordExtensibilitySamples/wiki/C-C---Custom-Visualizer-Sample).
```
Comment (gregg-miskelly) [inline CppCustomVisualizer/README.md]: The link should change to `https://github.com/Microsoft/ConcordExtensibilitySamples/wiki/Cpp-Custom-Visualizer-Sample`. GitHub wiki's apparently don't like '/' or '+' in the title.

File: CppCustomVisualizer/dll/_EntryPoint.cpp [hunk 2]
```diff
@@ -0,0 +1,371 @@
...
+
+    pBuffer += (cch-1); // '-1' is to convert from a character count (including null terminator) to a length
+    int remainaingLength = allocLength - (cch-1);
+
+    if (remainaingLength <= 1)
+        return HRESULT_FROM_WIN32(ERROR_INSUFFICIENT_BUFFER);
+
+    *pBuffer = ' ';
+    pBuffer++;
+    remainaingLength--;
```
Comment (rchiodo) [inline CppCustomVisualizer/dll/_EntryPoint.cpp]: What's this for? Adding a space?

Comment (gregg-miskelly) [reply CppCustomVisualizer/dll/_EntryPoint.cpp]: Yup

Comment (gregg-miskelly) [reply CppCustomVisualizer/dll/_EntryPoint.cpp]: I can add a comment.

File: CppCustomVisualizer/dll/_EntryPoint.cpp [hunk 3]
```diff
@@ -0,0 +1,371 @@
...
+#include "_EntryPoint.h"
+
+HRESULT STDMETHODCALLTYPE CCppCustomVisualizerService::EvaluateVisualizedExpression(
+    _In_ Evaluation::DkmVisualizedExpression* pVisualizedExpression,
+    _Deref_out_opt_ Evaluation::DkmEvaluationResult** ppResultObject
+    )
+{
+    HRESULT hr;
+
+    // This method is called to visualize the FILETIME variable. Its basic job is to create
```
Comment (gregg-miskelly) [inline CppCustomVisualizer/dll/_EntryPoint.cpp]: >the FILETIME variable [](start = 42, length = 21)

Should be "FILETIME variables"

Comment (gregg-miskelly) [reply CppCustomVisualizer/dll/_EntryPoint.cpp]: Or better 'a FILETIME variable'

---
In reply to: [251160417](https://github.com/Microsoft/ConcordExtensibilitySamples/pull/42#discussion_r251160417) [](ancestors = 251160417)

File: CppCustomVisualizer/dll/_EntryPoint.cpp [hunk 4]
```diff
@@ -0,0 +1,371 @@
...
+    DkmRootVisualizedExpression* pRootVisualizedExpression = DkmRootVisualizedExpression::TryCast(pVisualizedExpression);
+    if (pRootVisualizedExpression == nullptr)
+    {
+        // This sample doesn't provide child evaluation results, so only root expressions are expected
+        return E_NOTIMPL;
+    }
+
+    DkmInspectionContext* pParentInspectionContext = pVisualizedExpression->InspectionContext();
+
+    CComPtr<DkmLanguageExpression> pLanguageExpression;
```
Comment (gregg-miskelly) [inline CppCustomVisualizer/dll/_EntryPoint.cpp]: >CComPtr [](start = 4, length = 7)

I think this should actually be a CDkmAutoClosePtr

File: CppCustomVisualizer/vsix/source.extension.vsixmanifest [hunk 5]
```diff
@@ -0,0 +1,18 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<PackageManifest Version="2.0.0" xmlns="http://schemas.microsoft.com/developer/vsx-schema/2011" xmlns:d="http://schemas.microsoft.com/developer/vsx-schema-design/2011">
+  <Metadata>
+    <Identity Id="vsix..96d2707a-406e-423d-bbea-eb81c05d2b89" Version="1.0" Language="en-US" Publisher="Sample" />
```
Comment (gregg-miskelly) [inline CppCustomVisualizer/vsix/source.extension.vsixmanifest]: >vsix..96d2707a-406e-423d-bbea-eb81c05d2b89 [](start = 18, length = 42)

This likely needs to be updated

**Pull request merged**